### ISSUE TITLE
[🐛 bug report] NumPy conversion of Transform4f fails on the next branch

### ISSUE THREAD

**Issue by mehrab2603** [mitsuba-renderer/mitsuba2#400]
## Summary

Converting a `Transform4f` object to NumPy array on `cuda_ad_rgb` mode fails.

## System configuration

- Platform: Linux
- Compiler: Clang10
- Python version: 3.7
- Mitsuba 2 version: next branch (c238ddc75da15ad988421d799569503872d9d713)
- Compiled variants:
    * `scalar_rgb`
    * `scalar_spectral`
    * `cuda_rgb`
    * `cuda_ad_rgb`

## Description

Converting a `Transform4f` object to NumPy array on `cuda_ad_rgb` mode gives the following error (`torch()` doesn't work either btw).

> ValueError: cannot copy sequence with size 4 to array axis with dimension 1

Same code works fine in `scalar_rgb` mode.

## Steps to reproduce

- Set the mode to `cuda_ad_rgb`
- Do `scene.sensors()[0].world_transform().eval(3).matrix.numpy()`

**Comment by mehrab2603** [comment]
Hi @Speierers,
Seems like it was an issue with my NumPy version. Updating to 1.19 from 1.16 fixed this issue. Apologies for creating the issue. Closing.

**Comment by Speierers** [comment]
Regarding .torch() please pull latest next this should be fixed now.
Although I need to warn you, the pytorch integration in this new codebase isn't tested much for now.

**Comment by mehrab2603** [comment]
I am getting strange results in this next branch. The same code for creating a depthmap from world points given by the AOV integrator produces very different results in `scalar_rgb` and `cuda_ad_rgb` modes (both using `scene.integrator().render()` and not `autodiff.render_torch()`). The scalar_rgb results look correct to me.

**Comment by Speierers** [comment]
Hi @mehrab2603 ,

This works fine on my side, tested running the following command from the build directory:
```
mitsuba -m cuda_ad_rgb ../resources/data/docs/scenes/integrator_aov.xml 
```

Could you please share your script and scene so I can try to reproduce this bug? Maybe you could also share rendered images with both modes so we can understand what kind of "different results" it is.

Next time please open a seperate issue as it isn't related to the former request 😉

**Comment by mehrab2603** [comment]
Hi @Speierers,
These are the depthmaps I get from cuda_ad_rgb and scalar_rgb. The bottom one is scalar_rgb which gives me the expected results. 

![cuda_output](https://user-images.githubusercontent.com/5913653/111439051-9407ff80-86ca-11eb-86c7-7d8585e66003.png)
![scalar_output](https://user-images.githubusercontent.com/5913653/111439052-94a09600-86ca-11eb-9c6a-292f67cae67c.png)

Here is my code (needs PyTorch and OpenCV): [test.zip](https://github.com/mitsuba-renderer/mitsuba2/files/6154910/test.zip)

**Comment by Speierers** [comment]
Hi @mehrab2603 ,

This bug was related to the implementation of `__array_interface__` for the matrix types in Enoki and should now be fixed. Please pull `mitsuba/next` and update the submodules to give this another try.

Let me know if this fixes your problem.

**Comment by mehrab2603** [comment]
Sorry I could not test. I have pulled the latest commit on the next branch as of writing this (2232576cc50f45bb99873ef21423b67c79d34252), but I am getting a build error.

```
[build] [341/1041  31% :: 128.355] Building CXX object src/libcore/CMakeFiles/mitsuba-core.dir/string.cpp.o
[build] FAILED: src/libcore/CMakeFiles/mitsuba-core.dir/string.cpp.o 
[build] /usr/bin/clang++-10  -DLITTLE_ENDIAN -DMTS_BUILD_MODULE=MTS_MODULE_CORE -DMTS_ENABLE_AUTODIFF=1 -DMTS_ENABLE_CUDA=1 -DMTS_ENABLE_EMBREE=1 -DMTS_ENABLE_PROFILER=1 -DMTS_HAS_LIBPNG -D__LINUX__ -Dmitsuba_core_EXPORTS -I../include -I../ext/tinyformat -I../ext/enoki/include -I../ext/enoki/ext/enoki-jit/include -I../ext/enoki/ext/enoki-jit/ext/enoki-thread/include -Iinclude -I../ext/embree/include -I../ext/pugixml/src -I../ext/asmjit/src -I../ext/openexr/IlmBase/Imath -I../ext/openexr/IlmBase/Iex -I../ext/openexr/IlmBase/Half -I../ext/openexr/OpenEXR/IlmImf -Iext/openexr/OpenEXR/config -Iext/openexr/IlmBase/config -I../ext/openexr/IlmBase/IexMath -I../ext/openexr/IlmBase/IlmThread -stdlib=libc++ -D_LIBCPP_VERSION -march=native -fcolor-diagnostics -O3 -DNDEBUG -flto=thin -fPIC   -fno-math-errno -ffp-contract=fast -fno-trapping-math -fvisibility=hidden -Wdouble-promotion -Wall -Wextra -Wno-unused-local-typedefs -pthread -std=gnu++17 -MD -MT src/libcore/CMakeFiles/mitsuba-core.dir/string.cpp.o -MF src/libcore/CMakeFiles/mitsuba-core.dir/string.cpp.o.d -o src/libcore/CMakeFiles/mitsuba-core.dir/string.cpp.o -c ../src/libcore/string.cpp
[build] In file included from ../src/libcore/string.cpp:1:
[build] In file included from ../include/mitsuba/core/string.h:3:
[build] In file included from ../include/mitsuba/mitsuba.h:30:
[build] ../include/mitsuba/core/config.h:78:57: error: use of undeclared identifier 'CUDAArray'
[build]     else if constexpr (std::is_same_v<Float_, DiffArray<CUDAArray<float>>> &&
[build]                                                         ^
[build] ../include/mitsuba/core/config.h:78:74: error: expected a type
[build]     else if constexpr (std::is_same_v<Float_, DiffArray<CUDAArray<float>>> &&
[build]                                                                          ^
[build] ../include/mitsuba/core/config.h:78:76: error: type name requires a specifier or qualifier
[build]     else if constexpr (std::is_same_v<Float_, DiffArray<CUDAArray<float>>> &&
[build]                                                                            ^
[build] ../include/mitsuba/core/config.h:79:66: error: use of undeclared identifier 'CUDAArray'
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                                                                  ^
[build] ../include/mitsuba/core/config.h:79:83: error: expected a type
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                                                                                   ^
[build] ../include/mitsuba/core/config.h:79:88: error: expected a type
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                                                                                        ^
[build] ../include/mitsuba/core/config.h:79:29: error: expected unqualified-id
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                             ^
[build] ../include/mitsuba/core/config.h:81:5: error: expected expression
[build]     else
[build]     ^
[build] 8 errors generated.
[build] [341/1041  31% :: 129.088] Building CXX object ext/embree/kernels/CMakeFiles/embree_sse42.dir/geometry/curve_intersector_virtual2.cpp.sse42.cpp.o
[build] [341/1041  31% :: 129.177] Building CXX object src/libcore/CMakeFiles/mitsuba-core.dir/appender.cpp.o
[build] FAILED: src/libcore/CMakeFiles/mitsuba-core.dir/appender.cpp.o 
[build] /usr/bin/clang++-10  -DLITTLE_ENDIAN -DMTS_BUILD_MODULE=MTS_MODULE_CORE -DMTS_ENABLE_AUTODIFF=1 -DMTS_ENABLE_CUDA=1 -DMTS_ENABLE_EMBREE=1 -DMTS_ENABLE_PROFILER=1 -DMTS_HAS_LIBPNG -D__LINUX__ -Dmitsuba_core_EXPORTS -I../include -I../ext/tinyformat -I../ext/enoki/include -I../ext/enoki/ext/enoki-jit/include -I../ext/enoki/ext/enoki-jit/ext/enoki-thread/include -Iinclude -I../ext/embree/include -I../ext/pugixml/src -I../ext/asmjit/src -I../ext/openexr/IlmBase/Imath -I../ext/openexr/IlmBase/Iex -I../ext/openexr/IlmBase/Half -I../ext/openexr/OpenEXR/IlmImf -Iext/openexr/OpenEXR/config -Iext/openexr/IlmBase/config -I../ext/openexr/IlmBase/IexMath -I../ext/openexr/IlmBase/IlmThread -stdlib=libc++ -D_LIBCPP_VERSION -march=native -fcolor-diagnostics -O3 -DNDEBUG -flto=thin -fPIC   -fno-math-errno -ffp-contract=fast -fno-trapping-math -fvisibility=hidden -Wdouble-promotion -Wall -Wextra -Wno-unused-local-typedefs -pthread -std=gnu++17 -MD -MT src/libcore/CMakeFiles/mitsuba-core.dir/appender.cpp.o -MF src/libcore/CMakeFiles/mitsuba-core.dir/appender.cpp.o.d -o src/libcore/CMakeFiles/mitsuba-core.dir/appender.cpp.o -c ../src/libcore/appender.cpp
[build] In file included from ../src/libcore/appender.cpp:1:
[build] In file included from ../include/mitsuba/core/appender.h:3:
[build] In file included from ../include/mitsuba/core/object.h:5:
[build] In file included from ../include/mitsuba/core/class.h:3:
[build] In file included from ../include/mitsuba/mitsuba.h:30:
[build] ../include/mitsuba/core/config.h:78:57: error: use of undeclared identifier 'CUDAArray'
[build]     else if constexpr (std::is_same_v<Float_, DiffArray<CUDAArray<float>>> &&
[build]                                                         ^
[build] ../include/mitsuba/core/config.h:78:74: error: expected a type
[build]     else if constexpr (std::is_same_v<Float_, DiffArray<CUDAArray<float>>> &&
[build]                                                                          ^
[build] ../include/mitsuba/core/config.h:78:76: error: type name requires a specifier or qualifier
[build]     else if constexpr (std::is_same_v<Float_, DiffArray<CUDAArray<float>>> &&
[build]                                                                            ^
[build] ../include/mitsuba/core/config.h:79:66: error: use of undeclared identifier 'CUDAArray'
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                                                                  ^
[build] ../include/mitsuba/core/config.h:79:83: error: expected a type
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                                                                                   ^
[build] ../include/mitsuba/core/config.h:79:88: error: expected a type
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                                                                                        ^
[build] ../include/mitsuba/core/config.h:79:29: error: expected unqualified-id
[build]                        std::is_same_v<Spectrum_, Color<DiffArray<CUDAArray<float>>, 3>>)
[build]                             ^
[build] ../include/mitsuba/core/config.h:81:5: error: expected expression
[build]     else
[build]     ^
[build] 8 errors generated.
```

**Comment by wjakob** [comment]
did you run

```
git submodule update --init --recursive
```
?

**Comment by mehrab2603** [comment]
Yes, I actually did `git pullall` which is an alias for that.

**Comment by wjakob** [comment]
That's pretty weird then. What compiler and version are you using? What variants are enabled in your `mitsuba.conf`? What is the  revision of `next` that you have checked out?

**Comment by mehrab2603** [comment]
I am using clang 10. The variants are scalar_spectral scalar_rgb cuda_rgb and cuda_ad_rgb. And I have checked out 2232576cc50f45bb99873ef21423b67c79d34252. I'll try deleting everything and building from scratch.

**Comment by wjakob** [comment]
That would be great. Your setup sounds basically the same as what we are using, so it's weird that you are running into such a basic compiler error.

**Comment by mehrab2603** [comment]
I can build it successfully if I specifically clone the next branch (with git clone --single-branch --branch next --recursive https://github.com/mitsuba-renderer/mitsuba2). The compilation fails for some reason if I clone master, then checkout next. 

Anyway, I checked the resulting depthmap from both cuda_ad_rgb and scalar_rgb and they are same now. But I am getting another error now when I try to render my actual scene which is more complicated than the test scene I provided (`terminate called after throwing an instance of 'std::runtime_error' what():  jit_var_migrate(): Cannot resolve pointer to actual allocation!`) probably because of writing PyTorch tensors to param buffers. Code works fine if I don't modify params. I will probably open another issue after I investigate a bit more.
